src/S3/Transfer.php

```diff
@@ -233,18 +239,30 @@ private function createDownloadPromise()
             }
 
             // Create the command.
+            $args = $this->getS3Args($object);
+            $this->inProgress['Uncompleted'][] = "/" . $args['Bucket'] . "/" . $args['Key'];
             $commands []= $this->client->getCommand(
                 'GetObject',
-                $this->getS3Args($object) + ['@http'  => ['sink'  => $sink]]
+                $args + ['@http'  => ['sink'  => $sink]]
             );
         }
 
         // Create a GetObject command pool and return the promise.
         return (new Aws\CommandPool($this->client, $commands, [
             'concurrency' => $this->concurrency,
             'before'      => $this->before,
+            'fulfilled'   => function ($value, $idx, Promise\PromiseInterface $p) {
+                if ($value !== null && $value instanceof Aws\ResultInterface) {
+                    $uri = $value->get('@metadata')['effectiveUri'];
+                    $this->clearTaskDone((new Uri($uri))->getPath());
+                }
+            },
             'rejected'    => function ($reason, $idx, Promise\PromiseInterface $p) {
-                $p->reject($reason);
+                $cmd = $reason->getCommand();
+                $this->handleTransferException(
+                    $reason,
+                    "/" . $cmd['Bucket'] . "/" . $cmd['Key']
+                );
             }
         ]))->promise();
```

Comment:
This is where you'd wrap the exception by adding a `then` method on the `promise()` call that decorates the given exception, adds the information about the failed command, and the list of inFlightTransfers / remainingTransfers.